Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -378,6 +383,106 @@ def visit_AssertStatNode(self, node):
+            return names
+        # other targets are possible, but it isn't necessary to investigate them here
+        return []
+
+    def visit_ForInStatNode(self, node):
+        self.target_names[node] = tuple(self.find_target_names(node.target))
+        self.visitchildren(node)
+
+    def visit_ComprehensionNode(self, node):
+        pass  # don't recurse into nested comprehensions
+
+    def visit_LambdaNode(self, node):
+        pass  # don't recurse into nested lambdas/generator expressions
+
+    visit_Node = TreeVisitor.visitchildren
+
+
+class _AssignmentExpressionChecker(TreeVisitor):
+    """
+    Enforces rules on AssignmentExpressions within generator expressions and comprehensions
+    """
+    def __init__(self, loop_node, scope_is_class):
+        super(_AssignmentExpressionChecker, self).__init__()
+
+        target_name_finder = _AssignmentExpressionTargetNameFinder()
+        target_name_finder.visit(loop_node)
+        self.target_names_dict = target_name_finder.target_names
+
+        self.in_nested_generator = False
+        self.scope_is_class = scope_is_class
+        self.current_target_names = ()
+        self.all_target_names = set()
+        for names in self.target_names_dict.values():
+            self.all_target_names.update(names)
+
+    @classmethod
+    def do_checks(cls, loop_node, scope_is_class):
+        checker = cls(loop_node, scope_is_class)
+        checker.visit(loop_node)
+
+    def visit_ForInStatNode(self, node):
+        if self.in_nested_generator:
+            self.visitchildren(node)  # once nested, don't do anything special
+        else:
+            current_target_names = self.current_target_names
+            target_name = self.target_names_dict.get(node, None)
+            if target_name:
+                self.current_target_names += target_name
+
+            self.in_iterator = True
+            self.visit(node.iterator)
+            self.in_iterator = False
+            self.visitchildren(node, exclude=("iterator",))
+
+            self.current_target_names = current_target_names
+
+    def visit_AssignmentExpressionNode(self, node):
+        if self.in_iterator:
+            error(node.pos, "assignment expression cannot be used in a comprehension iterable expression")
+        if self.scope_is_class:
+            error(node.pos, "assignment expression within a comprehension cannot be used in a class body")
+        if node.target_name in self.current_target_names:
+            error(node.pos, "assignment expression cannot rebind comprehension iteration variable '%s'" %
+                  node.target_name)
+        elif node.target_name in self.all_target_names:
+            error(node.pos, "comprehension inner loop cannot rebind assignment expression target '%s'" %
+                  node.target_name)
+
+    def visit_LambdaNode(self, node):
+        # lambda node `def_node` isn't set up here
+        # so we need to recurse into it explicitly
+        self.visit(node.result_expr)
```

Comment:
Shouldn't `self.scope_is_class` be reset around this call? And what about the other flags, e.g. `self.in_iterator`?
Also, shouldn't the arguments to the lambda function override the assignment targets?
Might be worth a test or two.

For example, this is rejected by (plain) Python 3.10:
```python
>>> x = [ (y := z) for z in (lambda y: ([y, (a := y + 1)]))(2) ]
SyntaxError: assignment expression cannot be used in a comprehension iterable expression
```
Seems a current limitation that this PR appears to follow.

OTOH, the following works in Python:
```python
>>> y = 6
>>> x = (lambda y: (y := y + 1) + y)(2) + y
>>> x,y
(12, 6)
>>> class X:
...     x = (lambda y: (y := y + 1) + y)(2) + y
>>> X.x
12
```
And I think we'd currently reject it, right?

Reply:
> ```
>   >>> class X:
>    ...     x = (lambda y: (y := y + 1) + y)(2) + y
>    ```
> And I think we'd currently reject it, right?

No - that example works in both Python and Cython.

The issue is in assignment expressions in generators expressions in classes. An example of something that's rejected by Cython and accepted by Python is

```
class X:
    x = [(lambda y: (y := y + 1) + y)(2) + y for _ in range(2) ]
```

Which suggests that `visit_LambdaNode` should reset the flags. The next update will do that and and a few more tests. The tests copied from Python do contain some relevant test-cases too